// ngx-print/esm2022/lib/ngx-print.service.mjs
import { Injectable } from "@angular/core";
import { PrintBase } from "./ngx-print.base";
import * as i0 from "@angular/core";
/**
 * Service for handling printing functionality in Angular applications.
 * Extends the base printing class (PrintBase).
 *
 * @export
 * @class NgxPrintService
 * @extends {PrintBase}
 */
export class NgxPrintService extends PrintBase {
    /**
     * Initiates the printing process using the provided print options.
     *
     * @param {PrintOptions} printOptions - Options for configuring the printing process.
     * @memberof NgxPrintService
     * @returns {void}
     */
    print(printOptions) {
        // Call the print method in the parent class
        super.print(printOptions);
    }
    /**
     * Sets the print style for the printing process.
     *
     * @param {{ [key: string]: { [key: string]: string } }} values - A dictionary representing the print styles.
     * @memberof NgxPrintService
     * @setter
     */
    set printStyle(values) {
        super.setPrintStyle(values);
    }
    /**
     * Sets the stylesheet file for the printing process.
     *
     * @param {string} cssList - A string representing the path to the stylesheet file.
     * @memberof NgxPrintService
     * @setter
     */
    set styleSheetFile(cssList) {
        super.setStyleSheetFile(cssList);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: NgxPrintService, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: NgxPrintService, providedIn: "root" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: NgxPrintService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: "root",
                }]
        }] });
